platform/pc/boot/stage1.s

```diff
@@ -75,11 +75,42 @@ dap:
         .lba          dq 1
 
 
+%include "debug.inc"
+
 readsectors:
+        mov cx, [dap.sector_count]
+        mov ax, cx
+        mov dx, 0x0080
+        cmp cx, 0x0080
+        cmovnb cx, dx       ; cx = min(dap.sector_count, 0x80)
+        mov edx, 0x10000
+        sub edx, [dap.offset]
+        shr dx, 0x9
+        cmp dx, cx
+        cmovb cx, dx      ; cx = min(cx, (0x10000 - dap.offset) >> 0x9)
+        mov [dap.sector_count], cx
+loop:
         mov si, dap
         mov ah, 0x42
         mov dl, 0x80
         int 0x13
+        jc sector_read_error
+        sub ax, [dap.sector_count]
+        cmp ax, 0
+        je done 
```

Comment:
I found two issues here: as stated above, the segment needs to advance by 0x1000, as it is shifted left by 4 and added to the offset to make the effective address. Also, [dap.lba] needs to be advanced.

I was able to get the kernel to boot (with a >64kB stage2) using the following patch:

```
diff --git a/platform/pc/boot/stage1.s b/platform/pc/boot/stage1.s
index 3769fb69..2afc186f 100644
--- a/platform/pc/boot/stage1.s
+++ b/platform/pc/boot/stage1.s
@@ -97,10 +97,10 @@ loop:
         jc sector_read_error
         sub ax, [dap.sector_count]
         cmp ax, 0
-        je done 
-        mov dx, [dap.segment]
-        inc dx
-        mov [dap.segment], dx
+        je done
+        add [dap.lba], cx
+        mov cx, 0x1000
+        add [dap.segment], cx
         mov dx, 0x0
         mov [dap.offset], dx
         mov cx, 0x0080
```